index.js

```diff
@@ -1,120 +1,118 @@
 
-  while (lastProcessedPosition.blockNumber < lastConfirmedBlock) {
-    const toBlock = Math.min(lastProcessedPosition.blockNumber + constants.BLOCK_INTERVAL, lastConfirmedBlock)
+    if (lastPosition) {
+      this.lastProcessedPosition = lastPosition
+      logger.info(`Resuming export from position ${JSON.stringify(lastPosition)}`)
+    } else {
+      await this.exporter.savePosition(this.lastProcessedPosition)
+      logger.info(`Initialized exporter with initial position ${JSON.stringify(this.lastProcessedPosition)}`)
+    }
+  }
 
-    logger.info(`Fetching transfer events for interval ${lastProcessedPosition.blockNumber}:${toBlock}`)
-    metrics.requestsCounter.inc();
+  async workLoop() {
+    try {
+      const events = await this.worker.work()
+      metrics.currentBlock.set(this.worker.lastConfirmedBlock);
+      metrics.requestsCounter.inc();
+      metrics.requestsResponseTime.observe(new Date() - this.worker.lastRequestStartTime);
+      metrics.lastExportedBlock.set(this.worker.lastExportedBlock);
 
-    const requestStartTime = new Date();
-    let events = [];
-    if (constants.EXACT_CONTRACT_MODE) {
-      events = await getPastEventsExactContracts(web3, lastProcessedPosition.blockNumber + 1, toBlock);
-    }
-    else {
-      events = await getPastEvents(web3, lastProcessedPosition.blockNumber + 1, toBlock);
-    }
+      this.lastProcessedPosition.blockNumber = this.worker.lastExportedBlock
+      this.lastProcessedPosition.primaryKey = this.worker.lastPrimaryKey
 
-    metrics.requestsResponseTime.observe(new Date() - requestStartTime);
+      await storeEvents(this.exporter, events)
+      await this.exporter.savePosition(this.lastProcessedPosition)
+      logger.info(`Progressed to position ${JSON.stringify(this.lastProcessedPosition)}`)
 
-    if (events.length > 0) {
-      extendEventsWithPrimaryKey(events)
-      logger.info(`Storing and setting primary keys ${events.length} messages for blocks ${lastProcessedPosition.blockNumber + 1}:${toBlock}`)
-      await storeEvents(exporter, events)
-      lastProcessedPosition.primaryKey = events[events.length - 1].primaryKey
+      const _this = this
+      setTimeout(function() {_this.workLoop()}, _this.worker.sleepTimeMsec)
+    }
+    catch(ex) {
+      console.error("Error in exporter work loop: ", ex)
     }
-    lastExportTime = Date.now();
-
-    lastProcessedPosition.blockNumber = toBlock
-    metrics.lastExportedBlock.set(lastProcessedPosition.blockNumber);
-    await exporter.savePosition(lastProcessedPosition)
   }
 
-  logger.info(`Progressed to position ${JSON.stringify(lastProcessedPosition)}`)
-  // Look for new events every 30 sec
-  setTimeout(work, 30 * 1000)
-}
+  async setWorker() {
+    if (this.worker != null) {
+      throw new Error("Worker is already set")
+    }
 
-async function initLastProcessedBlock() {
-  const lastPosition = await exporter.getLastPosition()
+    switch (process.env.BLOCKCHAIN) {
```

Comment:
You can make worker initialization dynamic:
```js 
const WORKERS = { ERC20Worker, ERC20Worker }

export default function newWorker (name) {
  return WORKERS[name];
}
```
and then:
```js
const workerName = process.env.BLOCKCHAIN.toUpperCase()
this.worker = newWorker(workerName) 
```

Reply:
I am trying to weight in the pros of this syntax, do you find it cleaner? Also what do you mean by 'dynamic' am I missing some initialization difference?

Reply:
On a second thought - we won't add any new workers here in the future?

Reply:
The plan is to add new workers from other blockchains. I think I got your idea now, by using the syntax you propose one would need to include a new worker only as 'require' and not edit the code. Done similarly to how you proposed but with a 'new' operator.

Reply:
Yep, that was the idea.

Reply:
I've taken the idea up a notch. I dynamically use the blockchain name to load just the required file. This helps us not need contract mapping files for the ETH deploy for example.